File: glug_platform/src/macos/platform.m
#import "platform.h"
#import "platform_context.h"

#import <sys/utsname.h>
#import <stdio.h>
#import <Foundation/NSProcessInfo.h>

enum glug_os os_mac(const struct macos_context *context)
{
    (void) context;
    return glug_os_macos;
}

glug_bool os_version_mac(struct glug_plat_version *version, const struct macos_context *context)
{
    (void) context;
    NSOperatingSystemVersion os_vers = [[NSProcessInfo processInfo] operatingSystemVersion];
    version->major = (uint32_t)os_vers.majorVersion;
    version->minor = (uint32_t)os_vers.minorVersion;
    version->patch = (uint32_t)os_vers.patchVersion;

    return true;
}

glug_bool kernel_version_mac(struct glug_plat_version *version, const struct macos_context *context)
{
    (void) context;
    struct utsname utsname;
    uname(&utsname);

    sscanf(utsname.release, "%u.%u.%u", &version->major, &version->minor, &version->patch);

    return true;
}

glug_bool responds_to_osversion(void)
{
    return [NSProcessInfo instancesRespondToSelector: @selector(operatingSystemVersion)] == YES;
}

glug_bool os_version_fallback_mac(struct glug_plat_version *version, const struct macos_context *context)
{
    FILE *sw_vers = popen("sw_vers -productVersion", "r");

    if (!sw_vers) return false;

    fscanf(sw_vers, "%u.%u.%u", &version->major, &version->minor, &version->patch);
    pclose(sw_vers);

    return true;
}

File: glug_platform/src/platform/linux/platform.c
#include "platform.h"

#include <stdio.h>
#include <string.h>
#include <sys/utsname.h>

enum glug_os os_linux(void)
{
    return glug_os_linux;
}

void os_version_linux(struct glug_plat_version *version)
{
    (void) version;
}

void os_version_lsb(FILE *lsb_release, struct glug_plat_version *version)
{
    char tmp[20];
    int read = 0;

    fseek(lsb_release, 0, SEEK_SET);

    while((read = fscanf(lsb_release, " %20[^=]", tmp)))
    {
        if (!strcmp(tmp, "DISTRIB_RELEASE"))
        {
            fscanf(lsb_release, "=%u.%u.%u.%*u", &version->major, &version->minor, &version->patch);
            return;
        }
        // advance to next line
        fscanf(lsb_release, "%*[^\n]");
    }
}

void kernel_version_linux(FILE *proc_version, struct glug_plat_version *version)
{
    fseek(proc_version, 0, SEEK_SET);
    fscanf(proc_version, "%*[^0-9] %u.%u.%u", &version->major, &version->minor, &version->patch);
}

void kernel_version_uname(const uname_t uname, struct glug_plat_version *version)
{
    struct utsname utsname;

    if (uname(&utsname) == -1)
        return;

    sscanf(utsname.release, "%u.%u.%u", &version->major, &version->minor, &version->patch);
}
